docs/SILProgrammersManual.md

```diff
@@ -159,6 +159,165 @@ rules out PartialApplies is no excuse to conflate applied arguments
+involves a use-def traversal to determine the origin of the
+address. It may traverse operations on address, pointer, and reference
+types. The logic that formalizes which SIL operations may be involved
+in the def-use chain is encapsulated with the
+`AccessUseDefChainVisitor`. The traversal can be customized by
+implementing this visitor.
+
+Access def-use chains are divided by three points: the "root", the
+formal access "base", and the memory address.
+
+Def-use example:
+```
+  %root    = alloc_ref $C
+  %base    = ref_element_addr %root : $C, #C.prop
+  %access  = begin_access [read] [static] %base : $*S
+  %memaddr = struct_element_addr %access : $*S, #.field
+  %value   = load [trivial] %memaddr : $*Int64
+  end_access %access : $*S
+```
+
+The first part of the def-use chain computes the formal access base
+from the root of the object (e.g. alloc_ref -> ref_element_addr ->
+begin_access). Reference types are only allowed in this part. Each
+object property and its tail storage is considered a separate formal
+access base. Consequently the reference root is encapsulated within
+the `AccessedStorage` location, and apart from identifying the
+storage, projections from a reference type play no part in the
+AccessPath components.
+
+`AccessPath` extends `AccessedStorage` to also include the path
+components that determine the address of a memory operation within the
+formal access base (e.g. begin_access -> struct_element_addr ->
+load). Although the base of a formal access is always an address,
+conversions to and from pointers, as well as boxing and unboxing
+operations may still occur on this path.
+
+Phis may also occur on an access path, both on the reference root side
+and within the formal access, but only if each phi has no affect on
+the path components. Pointer casting and unboxing may occur on
+distinct phi paths, but index offsets and subobject projections may
+not. Address-type phis are prohibited. These rules are currently
+enforced to a limited extent, so it's possible for invalid access path
+to occur under certain conditions.
+
+Not all SIL memory access is part of a formal access, but the
+`AccessedStorage` and `AccessPath` abstractions are universally
+applicable. For non-formal access, certain invariants may not
+hold. For example, the access base may be a pointer type rather than
+an address type.
+
+The SILValue base of a formal access is recoverable from
+AccessedStorage in every case except for global variables. Global
+variable storage is uniquely identified, but it is impossible to find
+all uses from the def-use chain. It is possible to preserve an
+individual global variable base by using
+`AccessPathWithBase`. However, finding all uses still requires a
+separate algorithm for finding all global variable references within
+the function.
+
+### Access path uses
+
+For any accessed storage location and base, it must also be possible
+to reliably identify all uses of that storage location within the
+function.
```

Comment:
What if e.g. a function has references of the same class type as arguments? They might be aliased or not.
Or does collectUses() apply to a single begin/end_access scope?

Reply:
clarified